File: 01-fundamentals/server-python/src/mcp_fundamentals/hello.py
# ABOUTME: The simplest useful FastMCP server: one tool, one templated resource, one prompt.
# ABOUTME: Builder returns a FastMCP instance so tests can drive it with the in-memory client.
from __future__ import annotations

from fastmcp import FastMCP


def build_hello_server() -> FastMCP:
    """Build the hello server: an echo tool, a templated greeting resource, and a prompt."""
    mcp = FastMCP("fundamentals-hello")

    @mcp.tool
    def echo(text: str) -> str:
        """Return the given text unchanged. The smallest possible tool."""
        return text

    @mcp.resource("greeting://{name}")
    def greeting(name: str) -> str:
        """A templated resource: a greeting for the given name."""
        return f"Hello, {name}!"

    @mcp.prompt
    def summarize(topic: str) -> str:
        """A reusable prompt template asking for a short summary of a topic."""
        return f"Summarize the key points about {topic} in three bullets."

    return mcp

File: 01-fundamentals/server-typescript/src/hello.ts
// ABOUTME: TypeScript MCP hello server mirroring the Python one (echo tool, greeting resource, prompt).
// ABOUTME: Built on @modelcontextprotocol/sdk v1.x; buildHelloServer returns a connectable McpServer.
import { McpServer, ResourceTemplate } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

export function buildHelloServer(): McpServer {
  const server = new McpServer({ name: "fundamentals-hello-ts", version: "0.1.0" });

  server.registerTool(
    "echo",
    {
      title: "Echo",
      description: "Return the given text unchanged. The smallest possible tool.",
      inputSchema: { text: z.string() },
    },
    async ({ text }) => ({ content: [{ type: "text", text }] }),
  );

  server.registerResource(
    "greeting",
    new ResourceTemplate("greeting://{name}", { list: undefined }),
    { title: "Greeting", description: "A templated greeting resource." },
    async (uri, variables) => ({
      contents: [{ uri: uri.href, text: `Hello, ${String(variables.name)}!` }],
    }),
  );

  server.registerPrompt(
    "summarize",
    {
      title: "Summarize",
      description: "A reusable prompt asking for a short summary of a topic.",
      argsSchema: { topic: z.string() },
    },
    ({ topic }) => ({
      messages: [
        {
          role: "user",
          content: {
            type: "text",
            text: `Summarize the key points about ${topic} in three bullets.`,
          },
        },
      ],
    }),
  );

  return server;
}
